src/docs/arch-hs.dox

```diff
@@ -1,8 +1,123 @@
+
+@section arch_hs_cursor_visibilty History store cursor interface and visibility rules
+When using the history store cursor interface, the user can configure the type of visibility checks
+that are performed on the records. The behaviour is controlled by a set of cursor flags:
+
+- By default, snapshot based visibility checks are performed on all records returned to the API
+user.
+
+- When the flag \c WT_CURSTD_HS_READ_ALL is set on the cursor, no visibility checks are performed
+on the records returned to the API user. This means that cursor interface can even return a record
+with a globally visible tombstone. When this flag is set, it suppresses the effect of other
+visibility flags.
+
+- When the flag \c WT_CURSTD_HS_READ_COMMITTED is set, the cursor interface can return any record
+except for globally deleted records. This flag has no effect when the flag
+\c WT_CURSTD_HS_READ_ALL is set.
+
+History store cursor users must use one of the flags if the caller thread is running in a lower
+isolation level and doesn't hold a valid snapshot.
+
+@section arch_hs_reconciliation History store and reconciliation
+When a dirty page is reconciled on a user file b-tree, the update chain is examined and the latest
+committed update is chosen as the on-disk value. All older updates are added to the history store
+table assuming they are not yet obsolete. Additionally any out of order timestamps will
+be corrected.
+
+When a modified history store table page is processed by reconciliation, certain records on the page
+are removed when creating a new page image if the tombstones on those records have become globally
+visible. This ensures that WiredTiger is only keeping the relevant historical data required to
+serve the oldest reader in the system or as dictated by the oldest timestamp set by the application.
+Once all records on a page are obsolete, the page itself can be removed to reduce the size of
+the history store table (see @ref arch-transaction for more details about oldest timestamp).
+
+@section arch_hs_read Searching for older versions of a key in History Store
+When looking for an update visible to the current transaction, WiredTiger first searches the update
+chain for any visible updates. If there is no visible update in the chain, WiredTiger then
+inspects the on-disk version of the key. If that version is not visible to the transaction
+as well, WiredTiger searches the history store table for a visible update for the key. When
+searching for update that satisfies the read timestamp constraint, WiredTiger starts with the newest
+update of the key in the history store table and then iterates through the older updates until there
+are no updates left to process. Note that although there can be multiple updates in the
+history store for a key and read timestamp combination, each update would have different
+visibility based on transaction and timestamp based visibility rules.
```

Comment:
Should mention that WT_NOTFOUND will be returned to the searcher in the event none of the records are visible.

Reply:
Added this information.